azkaban-web-server/src/main/bash/azkaban-web-shutdown.sh

```diff
@@ -2,14 +2,33 @@
 # Shutdown script for azkaban web server
 installdir="$(dirname $0)/.."
 
+#currdir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+#installdir=$(dirname "${currdir}")
+maxtry=5
 pid=`cat ${installdir}/currentpid`
-echo "Killing Web Server. [pid: $pid]"
 
-kill ${pid}
-if [ $? -ne 0 ]; then
-    echo "Error: Shutdown failed"
-    exit 1;
+if [ -z $pid ]; then
+  echo "currentpid file doesn't exist in ${installdir}, shutdown completed"
+  exit 0
 fi
 
-rm  ${installdir}/currentpid
-echo "done."
+for try in $(seq 1 $maxtry); do
+  if [ ! -z $pid ]; then
+    echo "Killing Web Server. [pid: $pid], $try th try"
+    kill ${pid}
+    if [ -n "$(ps -p $pid -o pid=)" ]; then
+      echo "web server is not dead [pid: $pid]"
+      if [ $try -lt $maxtry ]; then
+        echo "sleeping for a few seconds before retry"
+        sleep 10 
+      fi
+    else 
+      rm  ${installdir}/currentpid
+      echo "shutdown succeeded"
+      exit 0
+    fi
+  fi
+done
+
+echo "Error: Shutdown failed"
+exit 1
```

Comment:
Yes, kill -9 would be an option. However, on balance I would rather warn the operator if the process can't be shut down and allow the operator to take manual actions. Otherwise, the failure will go unnoticed and it won't be much better than just do kill -9 always.